File: tools/analyze_unicode_key_ordering.js
"use strict";

/*
 * Investigate JavaScript ordering of Unicode composite identity keys.
 *
 * This intentionally compares a BMP private-use character with a
 * supplementary-plane emoji.
 *
 * Nothing in this module is normative OpenSmell.
 */

const PRIVATE_USE = "\uE000";
const EMOJI = "\u{1F600}";


function describe(value) {
    const codePoints = Array.from(value)
        .map(
            character =>
                "U+" +
                character
                    .codePointAt(0)
                    .toString(16)
                    .toUpperCase()
                    .padStart(4, "0")
        )
        .join(" ");

    return `${JSON.stringify(value)} (${codePoints})`;
}


function main() {
    console.log(
        "OpenSmell RFC-0006 JavaScript Unicode key-ordering experiment"
    );
    console.log("=".repeat(72));

    console.log();
    console.log("Keys");
    console.log("-".repeat(72));

    console.log(
        "Private-use:",
        describe(PRIVATE_USE)
    );

    console.log(
        "Emoji      :",
        describe(EMOJI)
    );

    const source = {
        [PRIVATE_USE]: "private-use",
        [EMOJI]: "emoji",
    };

    const sortedKeys = Object
        .keys(source)
        .sort();

    console.log();
    console.log("JavaScript sorted keys");
    console.log("-".repeat(72));

    for (const key of sortedKeys) {
        console.log(describe(key));
    }

    const orderedSource = {};

    for (const key of sortedKeys) {
        orderedSource[key] = source[key];
    }

    const canonical = JSON.stringify({
        dataset: "unicode_ordering",
        resource_type: "target",
        source_identity: orderedSource,
    });

    console.log();
    console.log("JavaScript canonical generation");
    console.log("-".repeat(72));

    console.log(canonical);

    console.log();
    console.log("UTF-8");
    console.log("-".repeat(72));

    console.log(
        Buffer
            .from(canonical, "utf8")
            .toString("hex")
    );
}


main();

File: tools/analyze_unicode_key_ordering.py
"""Investigate cross-language ordering of Unicode composite identity keys.

This experiment intentionally uses one BMP private-use character and one
supplementary-plane emoji to expose possible ordering differences between
Unicode code-point ordering and UTF-16 code-unit ordering.

Nothing in this module is normative OpenSmell.
"""

from __future__ import annotations

from opensmell.experimental.identifiers import canonical_generation_name


PRIVATE_USE = "\uE000"
EMOJI = "\U0001F600"


def describe(value: str) -> str:
    codepoints = " ".join(
        f"U+{ord(character):04X}"
        for character in value
    )
    return f"{value!r} ({codepoints})"


def main() -> None:
    print("OpenSmell RFC-0006 Unicode key-ordering experiment")
    print("=" * 72)

    print()
    print("Keys")
    print("-" * 72)
    print("Private-use:", describe(PRIVATE_USE))
    print("Emoji      :", describe(EMOJI))

    source_a = {
        PRIVATE_USE: "private-use",
        EMOJI: "emoji",
    }

    source_b = {
        EMOJI: "emoji",
        PRIVATE_USE: "private-use",
    }

    canonical_a = canonical_generation_name(
        dataset="unicode_ordering",
        resource_type="target",
        source_identity=source_a,
    )

    canonical_b = canonical_generation_name(
        dataset="unicode_ordering",
        resource_type="target",
        source_identity=source_b,
    )

    print()
    print("Python canonical generation")
    print("-" * 72)
    print(canonical_a)

    print()
    print("Insertion-order independence")
    print("-" * 72)
    print(
        "PASS"
        if canonical_a == canonical_b
        else "FAIL"
    )

    print()
    print("Python sorted keys")
    print("-" * 72)

    for key in sorted(source_a):
        print(describe(key))

    print()
    print("UTF-8")
    print("-" * 72)
    print(canonical_a.encode("utf-8").hex())


if __name__ == "__main__":
    main()